File: JavaScript/two_pointers/valid_palindrome.js
// https://leetcode.com/problems/valid-palindrome/

// Solution 1: Reverse The String and Compare
// Time Complexity: O(n)
// Space Complexity: O(n)

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  s = s.replace(/[^a-zA-Z0-9]/g, "").toLowerCase();
  return s.split("").reverse().join("") === s;
};

// Solution 2: Two Pointers with RegExp
// Time Complexity: O(n)
// Space Complexity: O(n) (if strings are immutable) / O(1) if strings are mutable

/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function (s) {
  s = s.replace(/[^a-zA-Z0-9]/g, "").toLowerCase();

  let start = 0;
  let end = s.length - 1;

  while (start <= end) {
    if (s[start] != s[end]) {
      return false;
    }
    start++;
    end--;
  }

  return true;
};

File: Python/two_pointers/valid_palindrome.py
# https://leetcode.com/problems/valid-palindrome/

class Solution:
    def isPalindrome(self, s: str) -> bool:
        newStr = ""

        for char in s:
            if char.isalnum():
                newStr += char.lower()

        return newStr == newStr[::-1]

    def isPalindrome2(self, s: str) -> bool:
        start, end = 0, len(s) - 1

        while start <= end:
            if not s[start].isalnum():
                start += 1
                continue

            if not s[end].isalnum():
                end -= 1
                continue

            if s[start].isalnum() and s[end].isalnum() and s[start].lower() != s[end].lower():
                return False

            start += 1
            end -= 1

        return True